### PR TITLE
AF-1097: [Project Oriented] Endless asset loading appears when discarding unsaved data

### PR DESCRIPTION
**Pull request by paulovmr**
Part of an ensemble:
* https://github.com/kiegroup/appformer/pull/293
* https://github.com/kiegroup/kie-wb-common/pull/1628
* https://github.com/kiegroup/drools-wb/pull/850
* https://github.com/kiegroup/jbpm-designer/pull/768

### CODE REVIEW

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java [hunk 1]
```diff
@@ -284,10 +292,26 @@ public void execute() {
...
+                    placeHistoryHandler.flush();
+                    launchPerspectiveActivity(place,
+                                              (PerspectiveActivity) activity,
+                                              doWhenFinished);
+                };
+
+                final PerspectiveActivity currentPerspective = perspectiveManager.getCurrentPerspective();
+
+                if (currentPerspective != null) {
+                    final BiParameterizedCommand<Command, PlaceRequest> closeChain = this.perspectiveCloseChain.get(currentPerspective.getIdentifier());
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java]: @paulovmr could you please add docs on the chain interface?

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java [hunk 2]
```diff
@@ -563,6 +597,20 @@ public boolean closeAllPlacesOrNothing() {
         return true;
     }
 
+    @Override
+    public List<PlaceRequest> getUncloseablePlaces() {
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java]: @paulovmr could you please add docs of the usage of this?

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java [hunk 3]
```diff
@@ -617,16 +665,27 @@ public void unregisterOnCloseCallbacks(final PlaceRequest place) {
         this.onCloseCallbacks.remove(place);
     }
 
+    @Override
+    public void registerPerspectiveCloseChain(final String perspectiveIdentifier,
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java]: docs

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java [hunk 4]
```diff
@@ -959,70 +1014,101 @@ private void openPartsRecursively(PanelDefinition panel) {
...
             return;
         }
 
-        workbenchPartBeforeCloseEvent.fire(new BeforeClosePlaceEvent(place,
-                                                                     force,
-                                                                     true));
-
-        closeSplashScreen(place);
-        activePopups.remove(place.getIdentifier());
+        final Command closeCommand = () -> {
```
Comment (ederign) [inline uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java]: please extract to a method

File: uberfire-api/src/main/java/org/uberfire/mvp/BiParameterizedCommand.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.uberfire.mvp;
+
+@FunctionalInterface
+public interface BiParameterizedCommand<T, U> {
```
Comment (ederign) [inline uberfire-api/src/main/java/org/uberfire/mvp/BiParameterizedCommand.java]: Docs ->

Paulo, this is not a bifunction in order to keep the relation with ParameterizedCommand<t> right? Maybe relate this with bifunction java interface just to help us understanding. Also, could you please refer this class ParametrizedCommand?

**Comment by ederign** [comment]
@paulovmr nice work. Just some comments. Can you please take a look?

Also, it's important to check the full downstream build with caution in this case. ;)

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java (hunk 1 above)
Comment (paulovmr) [reply uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java]: Done.

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java (hunk 2 above)
Comment (paulovmr) [reply uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java]: Added in the `PlaceManager` interface.

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java (hunk 3 above)
Comment (paulovmr) [reply uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java]: Added in the `PlaceManager` interface.

File: uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java (hunk 4 above)
Comment (paulovmr) [reply uberfire-workbench/uberfire-workbench-client/src/main/java/org/uberfire/client/mvp/PlaceManagerImpl.java]: Done.

File: uberfire-api/src/main/java/org/uberfire/mvp/BiParameterizedCommand.java (hunk 5 above)
Comment (paulovmr) [reply uberfire-api/src/main/java/org/uberfire/mvp/BiParameterizedCommand.java]: Docs done. Also referred to it in `ParameterizedCommand`.

**Pull request merged**